### PR TITLE
Enabled s3 `copyObject` for copy between s3 disks. 

### PR DESCRIPTION
**Pull request by MikhailBurdukov**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Made copy between s3 disks using a  s3-server-side copy instead of copying through the buffer. Improves `BACKUP/RESTORE`  operations and `clickhouse-disks copy` command.  

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Disks/ObjectStorages/DiskObjectStorage.cpp
Comment (CheSema) [inline src/Disks/ObjectStorages/DiskObjectStorage.cpp:191]: What has to be equal here to perform server side?
Could it be
- bucket?
- credentials?

What client should perform such requst?

PS: Types `s3` and `s3_plain` actually differ only in the way how CH maps files to blob keys.

Comment (MikhailBurdukov) [reply src/Disks/ObjectStorages/DiskObjectStorage.cpp:191]: >What has to be equal here to perform server side?

In fact, as I understood from  the doc([CopyObjectS3](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html)) and the current implementation of [`backup s3_policy to s3(...)`](https://github.com/ClickHouse/ClickHouse/blame/4ed4fbe74df0f48382bad2c55c883b763e5c869f/src/Backups/BackupIO_S3.cpp#L219) there are no any limitations on s3 side.  We limited it to the description of the data source so as not to handle scenarios when metadata_storage does not match. IMO, this may be more complicated than in the case with equal metadata_storage type. What is your opinion? Can we use server side with `s3` -> `s3_plain` scenario?

>What client should perform such requst?

Do you mean client for source s3 or destination? Seems like client must be to destination object_storage similar like [here](https://github.com/ClickHouse/ClickHouse/blob/4ed4fbe74df0f48382bad2c55c883b763e5c869f/src/Backups/BackupIO_S3.cpp#L197). But right now requests for copy object are sending to source s3, will fix it.

Comment (CheSema) [reply src/Disks/ObjectStorages/DiskObjectStorage.cpp:193]: > Do you mean client for source s3 or destination? Seems like client must be to destination object_storage similar like [here](https://github.com/ClickHouse/ClickHouse/blob/4ed4fbe74df0f48382bad2c55c883b763e5c869f/src/Backups/BackupIO_S3.cpp#L197). But right now requests for copy object are sending to source s3, will fix it.

Yes I agree, it is natural to get the dst-client for such operation. Because that client needs to have only additional RO access to other disk. Otherwise the client needs to have write access to other disk.

Comment (CheSema) [reply src/Disks/ObjectStorages/DiskObjectStorage.cpp:193]: > Can we use server side with s3 -> s3_plain scenario?
You do not have to solve this actually here. I dont know how to solve it either. That entities are pure logical. That decision depends on its logic. If I have to do it, I would revise the reasons why that types of disks even exists.

Comment (MikhailBurdukov) [reply src/Disks/ObjectStorages/DiskObjectStorage.cpp:193]: @CheSema Hi! Could you please check the PR?

Comment (MikhailBurdukov) [reply src/Disks/ObjectStorages/DiskObjectStorage.cpp:193]: @CheSema ping

File: src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp
Comment (CheSema) [inline src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:485]:
```suggestion
            if (exc.getS3ErrorCode() != Aws::S3::S3Errors::ACCESS_DENIED)
                throw;
                
```

File: src/Disks/ObjectStorages/DiskObjectStorage.cpp
Comment (MikhailBurdukov) [reply src/Disks/ObjectStorages/DiskObjectStorage.cpp:193]: >Do we perform server side copy for different buckets?

Yes. as I can see we're already performing native copy in scenario: `backup .. to s3`([the test](https://github.com/ClickHouse/ClickHouse/blob/e04416b48f20037ec2da5deb411254d3c993592e/tests/integration/test_backup_restore_s3/test.py#L317))

**Review by CheSema — APPROVED** [review]

**Comment by CheSema** [comment]
There are a lot of failing tests. I just can't merge it until they resolved.

**Comment by MikhailBurdukov** [comment]
Most of them look like this:
```
E           subprocess.CalledProcessError: Command '['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-m', 'statistic', '--mode', 'random', '--probability', '0.5', '-p', 'tcp', '-s', '172.16.1.5', '--dport', '2181', '-j', 'DROP']' returned non-zero exit status 126.
```
IMO its strange, will restart the CI. Can't  reproduce it locally.

**Comment by CheSema** [comment]
Please take a look at
test_storage_azure_blob_storage/test_cluster.py::test_unset_skip_unavailable_shards
That test seemed to be stable

https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksICBncm91cFVuaXFBcnJheShwdWxsX3JlcXVlc3RfbnVtYmVyKSwgIGFueShyZXBvcnRfdXJsKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMy0wMS0wMScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X3Vuc2V0X3NraXBfdW5hdmFpbGFibGVfc2hhcmRzJScgYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScpIGdyb3VwIGJ5IGQgb3JkZXIgYnkgZCBkZXNj

**Comment by MikhailBurdukov** [comment]
As I can see the test shouldn't execute my changes at all:
```
    node = cluster.instances["node_0"]
    storage_account_url = cluster.env_variables["AZURITE_STORAGE_ACCOUNT_URL"]
    azure_query(
        node,
        f"INSERT INTO TABLE FUNCTION azureBlobStorage('{storage_account_url}', 'cont', 'test_skip_unavailable.csv', 'devstoreaccount1', "
        f"'Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==', 'auto', "
        f"'auto', 'a UInt64') VALUES (1), (2)",
    )
    result = azure_query(
        node,
        f"SELECT count(*) from azureBlobStorageCluster('cluster_non_existent_port','{storage_account_url}', 'cont', 'test_skip_unavailable.csv', "
        f"'devstoreaccount1','Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==') "
        f"SETTINGS skip_unavailable_shards = 1",
    )
```
Also I can't reproduce the result on the local machine.

CI output:
```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 23.12.1):
E           Code: 1001. DB::Exception: Received from 172.16.10.5:9000. DB::Exception: Azure::Storage::StorageException: 404 The specified blob does not exist.
E           
E           Request ID: df9645f0-6c8a-40d6-958d-18c53991491b. Stack trace:
E           
E           0. ./contrib/llvm-project/libcxx/include/exception:134: std::runtime_error::runtime_error(String const&) @ 0x000000004079418e in /usr/bin/clickhouse
E           1. ./contrib/azure/sdk/core/azure-core/inc/azure/core/exception.hpp:29: Azure::Storage::StorageException::CreateFromResponse(std::unique_ptr<Azure::Core::Http::RawResponse, std::default_delete<Azure::Core::Http::RawResponse>>) @ 0x0000000034ddbe6f in /usr/bin/clickhouse
E           2. ./contrib/azure/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp:7198: Azure::Storage::Blobs::_detail::BlobRestClient::Blob::GetProperties(Azure::Core::Http::_internal::HttpPipeline&, Azure::Core::Url const&, Azure::Storage::Blobs::_detail::BlobRestClient::Blob::GetBlobPropertiesOptions const&, Azure::Core::Context const&) @ 0x0000000034d3d73f in /usr/bin/clickhouse
E           3. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: Azure::Storage::Blobs::BlobClient::GetProperties(Azure::Storage::Blobs::GetBlobPropertiesOptions const&, Azure::Core::Context const&) const @ 0x0000000034d2f0db in /usr/bin/clickhouse
E           4. ./build_docker/./src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:0: DB::AzureObjectStorage::getObjectMetadata(String const&) const @ 0x000000002b4c7b05 in /usr/bin/clickhouse
E           5. ./build_docker/./src/Storages/StorageAzureBlob.cpp:0: DB::StorageAzureBlobSource::KeysIterator::KeysIterator(DB::AzureObjectStorage*, String const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::IAST>, DB::NamesAndTypesList const&, std::shared_ptr<DB::Context const>, std::vector<DB::RelativePathWithMetadata, std::allocator<DB::RelativePathWithMetadata>>*, std::function<void (DB::FileProgress)>) @ 0x0000000030942094 in /usr/bin/clickhouse
E           6. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: std::__shared_ptr_emplace<DB::StorageAzureBlobSource::KeysIterator, std::allocator<DB::StorageAzureBlobSource::KeysIterator>>::__shared_ptr_emplace[abi:v15000]<DB::AzureObjectStorage*&, String const&, std::vector<String, std::allocator<String>> const&, std::nullptr_t, DB::NamesAndTypesList, std::shared_ptr<DB::Context const>&, std::vector<DB::RelativePathWithMetadata, std::allocator<DB::RelativePathWithMetadata>>*>(std::allocator<DB::StorageAzureBlobSource::KeysIterator>, DB::AzureObjectStorage*&, String const&, std::vector<String, std::allocator<String>> const&, std::nullptr_t&&, DB::NamesAndTypesList&&, std::shared_ptr<DB::Context const>&, std::vector<DB::RelativePathWithMetadata, std::allocator<DB::RelativePathWithMetadata>>*&&) @ 0x000000003097e469 in /usr/bin/clickhouse
E           7. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:953: DB::StorageAzureBlob::getTableStructureFromData(DB::AzureObjectStorage*, DB::StorageAzureBlob::Configuration const&, std::optional<DB::FormatSettings> const&, std::shared_ptr<DB::Context const>, bool) @ 0x0000000030932119 in /usr/bin/clickhouse
E           8. ./build_docker/./src/Storages/StorageAzureBlobCluster.cpp:51: DB::StorageAzureBlobCluster::StorageAzureBlobCluster(String const&, DB::StorageAzureBlob::Configuration const&, std::unique_ptr<DB::AzureObjectStorage, std::default_delete<DB::AzureObjectStorage>>&&, DB::StorageID const&, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, std::shared_ptr<DB::Context const>, bool) @ 0x000000003097f564 in /usr/bin/clickhouse
E           9. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: std::shared_ptr<DB::StorageAzureBlobCluster> std::allocate_shared[abi:v15000]<DB::StorageAzureBlobCluster, std::allocator<DB::StorageAzureBlobCluster>, String const&, DB::StorageAzureBlob::Configuration&, std::unique_ptr<DB::AzureObjectStorage, std::default_delete<DB::AzureObjectStorage>>, DB::StorageID, DB::ColumnsDescription&, DB::ConstraintsDescription, std::shared_ptr<DB::Context const>&, bool&, void>(std::allocator<DB::StorageAzureBlobCluster> const&, String const&, DB::StorageAzureBlob::Configuration&, std::unique_ptr<DB::AzureObjectStorage, std::default_delete<DB::AzureObjectStorage>>&&, DB::StorageID&&, DB::ColumnsDescription&, DB::ConstraintsDescription&&, std::shared_ptr<DB::Context const>&, bool&) @ 0x000000002997f562 in /usr/bin/clickhouse
E           10. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962: DB::TableFunctionAzureBlobStorageCluster::executeImpl(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool) const @ 0x000000002997b497 in /usr/bin/clickhouse
E           11. ./build_docker/./src/TableFunctions/ITableFunction.cpp:37: DB::ITableFunction::execute(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool, bool) const @ 0x000000002b0c866b in /usr/bin/clickhouse
E           12. ./build_docker/./src/Interpreters/Context.cpp:1850: DB::Context::executeTableFunction(std::shared_ptr<DB::IAST> const&, DB::ASTSelectQuery const*) @ 0x000000002d1b138c in /usr/bin/clickhouse
E           13. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::JoinedTables::getLeftTableStorage() @ 0x000000002f20a51c in /usr/bin/clickhouse
E           14. ./build_docker/./src/Interpreters/InterpreterSelectQuery.cpp:0: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>) @ 0x000000002ef22087 in /usr/bin/clickhouse
E           15. ./build_docker/./src/Interpreters/InterpreterSelectQuery.cpp:209: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x000000002ef1f0b7 in /usr/bin/clickhouse
E           16. ./build_docker/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:257: DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::shared_ptr<DB::IAST> const&, std::vector<String, std::allocator<String>> const&) @ 0x000000002f12f558 in /usr/bin/clickhouse
E           17. ./build_docker/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:0: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x000000002f12b41a in /usr/bin/clickhouse
E           18. ./contrib/llvm-project/libcxx/include/vector:580: std::__unique_if<DB::InterpreterSelectWithUnionQuery>::__unique_single std::make_unique[abi:v15000]<DB::InterpreterSelectWithUnionQuery, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&>(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&) @ 0x000000002ee49ef4 in /usr/bin/clickhouse
E           19. ./build_docker/./src/Interpreters/InterpreterFactory.cpp:0: DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x000000002ee46eaf in /usr/bin/clickhouse
E           20. ./build_docker/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000002fbf0565 in /usr/bin/clickhouse
E           21. ./build_docker/./src/Interpreters/executeQuery.cpp:1287: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x000000002fbe81ca in /usr/bin/clickhouse
E           22. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:612: DB::TCPHandler::runImpl() @ 0x0000000032b24ce9 in /usr/bin/clickhouse
E           23. ./build_docker/./src/Server/TCPHandler.cpp:2294: DB::TCPHandler::run() @ 0x0000000032b69490 in /usr/bin/clickhouse
E           24. ./build_docker/./base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x000000003a8497cf in /usr/bin/clickhouse
E           25. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: Poco::Net::TCPServerDispatcher::run() @ 0x000000003a84a4f7 in /usr/bin/clickhouse
E           26. ./build_docker/./base/poco/Foundation/src/ThreadPool.cpp:0: Poco::PooledThread::run() @ 0x000000003ac55edc in /usr/bin/clickhouse
E           27. ./base/poco/Foundation/src/Thread_POSIX.cpp:0: Poco::ThreadImpl::runnableEntry(void*) @ 0x000000003ac4eea8 in /usr/bin/clickhouse
E           28. asan_thread_start(void*) @ 0x000000000a84525b in /usr/bin/clickhouse
E           29. ? @ 0x00007f46390e7ac3 in ?
E           30. ? @ 0x00007f4639179a40 in ?
E           . (STD_EXCEPTION)
E           (query: SELECT count(*) from azureBlobStorageCluster('cluster_non_existent_port','http://azurite1:61405/devstoreaccount1', 'cont', 'test_skip_unavailable.csv', 'devstoreaccount1','Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw=='))
```

The same results in others PR's 
- https://s3.amazonaws.com/clickhouse-test-reports/56064/24e1e65dc0eb9187de45a74020fb4c21252d90ea/integration_tests__release__[4_4].html
- Failure on mater build: https://s3.amazonaws.com/clickhouse-test-reports/0/7a5b40563ab6625e2e2722cd8ba3fa282e82037c/integration_tests__release__[4_4].html 
- https://s3.amazonaws.com/clickhouse-test-reports/57793/eff4196be16a454d7f5c84721d3b83c7d52d659e/integration_tests__asan__[4_4].html

**Comment by flozdra** [comment]
Hello,

As far as I understand, the s3-server-side copy is possible between a disk `s3` and `s3_plain`?

If so, then the documentation led me wrong as it states : 
> If your tables are backed by S3 storage and **types of the disks are different**, it doesn't use `CopyObject` calls to copy parts to the destination bucket, instead, it downloads and uploads them, which is very inefficient. Prefer to use `BACKUP ... TO S3(<endpoint>)` syntax for this use-case.

I tried to setup a `MergeTree` table backed in S3 using a `s3` disk, and then perform a `BACKUP` of this table on a `s3_plain` disk, and the command uses s3-server-side copy with `CopyObject` calls.

Am I missing something?

**Comment by azat** [comment]
>Am I missing something?

Nope, and AFAICS it should work, can you send a PR to update the documentation?

**Comment by MikhailBurdukov** [comment]
yes, by now it should works. 

[Previously](https://github.com/MikhailBurdukov/ClickHouse/blob/119e4519675da22d8516e384cda41f9d4a3e9c11/src/Disks/DiskType.h#L9C1-L19C3), disks did not have  separation for storage type/metadata type. After refactoring(https://github.com/ClickHouse/ClickHouse/pull/58963), a check for the metadata type was not added in the [comparison operator](https://github.com/ClickHouse/ClickHouse/blob/843e9db8d17fe30f7f977622bf969564eed0b4d1/src/Disks/DiskType.cpp#L28). 
Suggesting that  unexpected consequences made native copying between plain/local disks possible. :)

**Pull request merged**
